`BundleMaker` REST call allowing to add ZIP/JAR to catalog - Part #2

tbouron opened the pull request:
Builds on top of #485. It implements the missing bits and pieces, such as:
- fields `version` and `bundle` are required under `brooklyn.catalog` when uploading a ZIP/JAR archive
- returns 400 HTTP status code is anything goes wrong during the process of the archive (as this is a bad input, therefore a bad request)
- avoid using the `CatalogPopulator` as it reloads all bundles. Instead, it uses the new `CatalogBundleLoader` that loads only a given bundle
- if the bundle has been installed, but the CAMP parser fails, the bundle is uninstalled
- adds more unit tests, especially to check error handling

geomacy commented on core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java:
```diff
@@ -195,22 +210,29 @@ public void removedBundle(Bundle bundle, BundleEvent bundleEvent, Iterable<? ext
-                }, " "), e);
-            }
+    @Beta
+    public class CatalogBundleLoader {
```
I'd say it's worth splitting this out into a standalone class. Same for CatalogPopulator; which I think we should rename to something like 'CatalogLoadingBundleTracker'.

geomacy replied:
In CatalogBomScanner, now that this refactorization has being done we should remove the constructor 
```
CatalogPopulator(BundleContext bundleContext, ManagementContext managementContext)
``` 
which was introduced for this purpose, including deleting the `bundleContext` and `managementContext` members and re-simplifying any methods that use them.

geomacy replied:
We should also remove the changes introduced in 33f7940c18ba8cd17fd83500486e02376d3b8fd8 to CatalogPopulator, i.e. adding the `CatalogPopulator(BundleContext bundleContext, ManagementContext managementContext) ` constructor and `bundleContext` member variable, and re-simplifying `open()` and `close()` and anything else that uses them.  Those changes were introduced for the purposes of this PR, but are not needed now that the new class has been factored out.

geomacy commented on rest/rest-api/src/main/java/org/apache/brooklyn/rest/api/CatalogApi.java:
```diff
@@ -38,27 +38,60 @@
+    @ApiOperation(value = "Add a catalog item (e.g. new type of entity, policy or location) by uploading OSGi bundle JAR, or ZIP which will be turned into bundle JAR, "
+            + "containing catalog.bom containing bundle name and version. "
+            + "Return value is 201 CREATED if bundle could be added.", response = String.class)
+    public Response createFromArchive(
```
Would it be worth deleting this and `createFromYaml` so that Swagger on the UI had only one endpoint to present?

geomacy commented on rest/rest-resources/src/main/java/org/apache/brooklyn/rest/resources/CatalogResource.java:
```diff
@@ -129,6 +176,97 @@ public Response create(String yaml) {
+            if (!BrooklynFeatureEnablement.isEnabled(BrooklynFeatureEnablement.FEATURE_LOAD_BUNDLE_CATALOG_BOM)) {
+                // if the above feature is not enabled, let's do it manually (as a contract of this method)
+                try {
+                    new CatalogBomScanner().new CatalogBundleLoader(mgmt()).scanForCatalog(bundle);
```
This should just be `new CatalogBundleLoader(mgmt()).scanForCatalog(bundle);` (with it as a standalone class)

tbouron commented:
@geomacy I addressed your comments and separated out the `CatalogBomScanner`, `CatalogBundleTracker` and `CatalogBundleLoader`. I also dug into the swagger docs and found out how to hide specific `@ApiOperation`. I had to bump the version of the `swagger-maven-plugin` because it was throwing an NPE with the new `hidden` parameter.

I did a full build locally, with the latest master of the other submodules, everything went fine so I'm guessing the error reported by Jenkins was due to something else

neykov commented:
@tbouron Looks like #485 broke the `brooklyn-client` build, can you try fixing it here.
See https://builds.apache.org/view/Brooklyn/job/brooklyn-client-master/41/console.

tbouron commented:
@neykov Yep, I'll look into it

neykov commented:
Thanks @tbouron.

tbouron commented:
@neykov Should be fixed now. I also created https://github.com/apache/brooklyn-client/pull/41 to use the non-deprecated call

geomacy commented:
hi @tbouron I get an error trying to build this
```
[ERROR] Failed to execute goal com.github.kongchen:swagger-maven-plugin:3.1.4:generate (default) on project brooklyn-rest-api: Unable to parse configuration of mojo com.github.kongchen:swagger-maven-plugin:3.1.4:generate for parameter locations: Cannot assign configuration entry 'locations' with value 'org.apache.brooklyn.rest.api' of type java.lang.String to property of type java.util.List -> [Help 1]
```
I just synced and built this as-is, should I be merging pre-requisite PRs or something?

tbouron commented:
@geomacy Hum odd. I have build this multiple time locally with the latest master of all the sub-modules. Jenkins also seems happy.

I did [another PR](https://github.com/apache/brooklyn-client/pull/41) in `brooklyn-client` to fix the tests in the client (related to the changes I did here) but I don't see how that could be related.

geomacy commented:
aha - Maven version issue.   Upgrading 3.3.3 -> 3.3.9 fixed the problem.

neykov commented:
Sounds like we should bump the maven version requirement - https://github.com/apache/brooklyn-server/blob/master/parent/pom.xml#L948. Or perhaps downgrade `swagger-maven-plugin`?

tbouron commented:
@neykov we cannot downgrade `swagger-maven-plugin` because of the issue I reported in [my comment](https://github.com/apache/brooklyn-server/pull/608#issuecomment-289000101).

We can either update the plugin configuration to properly use list instead of comma separated string or bump the maven requirement. Up to you.

neykov commented:
Not sure what you mean by "update the plugin configuration". Couldn't see any related configuration in the pom.xml. No strong opinion though, either one is fine by me.

tbouron commented:
@neykov I'm talking [about this](https://github.com/apache/brooklyn-server/pull/608/commits/5fb40ff8ae077da061aee927a773ec7a232df97d)

neykov commented:
Thanks for the pointer, missed that in the changes.

tbouron commented:
@neykov You didn't miss it, I added it after your comment to solve the issue ;)

geomacy commented on rest/rest-resources/src/main/java/org/apache/brooklyn/rest/resources/CatalogResource.java:
```diff
@@ -245,13 +246,21 @@ public Response createFromArchive(byte[] zipInput) {
-                        ((LocalManagementContext) mgmt()).getOsgiManager().get().getFramework().getBundleContext(),
-                        mgmt()).addingBundle(bundle, null);
+                try {
+                    new CatalogBundleLoader(new CatalogBomScanner(), mgmt()).scanForCatalog(bundle);
```
This is a bit of a problem; you shouldn't have to pass a `CatalogBomScanner()` in here; the point of doing this decoupling was so that the `CatalogBundleLoader` could avoid knowing about the `CatalogBomScanner`. 

I see that the use that is made of this is handling the [whitelist/blacklist](https://github.com/apache/brooklyn-server/commit/7046ba59cefd7e82ffad3f97057b3646915d1f08#diff-c46e6ef1f0a91ac1f58e7a340ecc0f34) functionality.  (Well there is also `CatalogBomScanner.bundleIds()` but that can be made a static method,
and maybe moved to `CatalogUtils`.)  

This whitelist stuff is configured [here](https://github.com/apache/brooklyn-server/blob/master/karaf/init/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L123), in the definition of the singleton scanner object. (Only supported in Karaf launcher at the moment, as you see.)

Doing `new CatalogBundleLoader(new CatalogBomScanner(), mgmt()).scanForCatalog(bundle);` means you are supplying an object that hasn't been configured with the right white/blacklists, so it bypasses this restriction. 

I'm going to suggest this is something we can live with for now, but mark as a "TODO" for future action. 

However, it is still ugly to have the dependency on the `CatalogBomScanner` in the tracker and the loader, so I suggest we replace it with a simple predicate that wraps the whitelist checks and inject the predicate into the loader. That way the CatalogBomScanner can inject the correct whitelist configuration in the blueprint.xml, while in the REST API here we just live with an "always true" for the moment. 

The loader could also be created in the scanner and injected into the tracker, which will simplify the latter a bit.

That's all a bit vague; I have to play around with the code a bit to help me think, so here's a PR that 
illustrates what I mean - https://github.com/tbouron/brooklyn-server/pull/2

What do you think?

tbouron replied:
@geomacy Yup, the dependency on `CatalogBomScanner` is ugly. I knew it had to be changed but wanted this to be merged as soon as possible so I just moved the code around rather than really refactor.

Re the whitelist injection for the REST API: I was wondering what if we add the `CatalogBomScanner` to the new `ScratchPad` area? It's supposed to be a singleton isn't it? We can then get back the correctly setup predicate from it.

WDYT?

geomacy commented:
I hope that PR of mine could just be used as-is to finish off that refactoring.  Personally I wouldn't worry too much about getting this integrated just now with a singleton `CatalogBomScanner`, especially as this is disabled by default at the minute anyway (see #233), until it can be made to work well with rebind.

geomacy commented:
I think this looks good; further work as discussed earlier can be done as a separate PR. Will merge.

neykov commented:
@tbouron started getting test failures in [VanillaSoftwareProcessTest.testVanillaSoftwareProcessCanUseResourcesInBundles](https://builds.apache.org/view/Brooklyn/job/brooklyn-integration-tests/70/testngreports/org.apache.brooklyn.rest.test.entity/VanillaSoftwareProcessTest/testVanillaSoftwareProcessCanUseResourcesInBundles/) around the catalog changes. Could you have a look.

tbouron commented:
@neykov Fixed in #608

neykov commented:
That was quick, thanks @tbouron.

aledsage commented on core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBundleLoader.java:
```diff
@@ -0,0 +1,180 @@
+     * @return A list of items added to the catalog
+     * @throws RuntimeException if the catalog items failed to be added to the catalog
+     */
+    public Iterable<? extends CatalogItem<?, ?>> scanForCatalog(Bundle bundle) {
```
I'd prefer a name that makes clear the method will add the catalog items (as well as scanning + returning the items found). For example `addCatalogItemsFromBundle(...)`?

(This method name was originally in `CatalogBomScanner`, and was refactored to here - but in CatalogBomScanner it was a private method so the name didn't need to be as clear.)

ahgittin commented:
Thanks for polishing this off @tbouron !
